text/0745-user-provided-cdk-init-templates.md

```diff
@@ -0,0 +1,1628 @@
+# RFC: User-Provided `cdk init` Templates
+
+Original Author(s):: @rohang9000
+
+Tracking Issue: [#745](https://github.com/aws/aws-cdk-rfcs/issues/745)
+
+API Bar Raiser: @iliapolo
+
+Customers need a way to initialize CDK projects in the CLI with custom templates to meet specific technical/regulatory requirements and standardize project initialization.
+
+
+## Working Backwards
+
+This section of the RFC addresses both users who will use custom templates to initialize a CDK project through the `cdk init` command and authors of custom templates.
+
+
+### CHANGELOG
+
+```
+feat(cli): user-provided cdk init templates
+```
+
+
+
+### Updated README for Users of CDK CLI
+
+#### `cdk init`
+
+Creates a new CDK project from built-in templates or custom templates. This command helps you get started with AWS CDK by initializing a new project with the necessary files and folders to organize your CDK code, including directories for your application definition, stacks, and constructs.
+
+
+#### Getting Started with Built-in Templates
+
+The CDK provides three built-in templates to help you get started quickly:
+
+1. `app` - CDK Application Template
+
+    * Creates a basic deployable CDK application structure
+    * Includes an empty stack
+    * Available in: TypeScript, JavaScript, Python, Java, C#, F#, Go
+
+2. `lib` - CDK Construct Library Template
+
+    * For building and sharing custom CDK constructs
+    * Available in: TypeScript
+
+3. `sample-app` - Example CDK Application
+
+    * Creates a CDK application with example constructs
+    * Demonstrates common AWS services and patterns
+    * Available in: TypeScript, JavaScript, Python, Java, C#, F#, Go
+
+##### Using Built-In Templates
+
+```
+# Syntax
+# Use a template for a CDK Application
+$ cdk init app --language=[csharp|fsharp|go|java|javascript|python|typescript]
+
+# Use a template for a CDK Construct Library
+$ cdk init lib --language=typescript
+
+# Use a template for an example CDK Application with some constructs
+$ cdk init sample-app --language=[csharp|fsharp|go|java|javascript|python|typescript]
+
+# Examples:
+# Create a new TypeScript application
+$ cdk init app --language=typescript
+
+# Create a construct library
+$ cdk init lib --language=typescript
+
+# Create a sample TypeScript application to explore CDK features
+$ cdk init sample-app --language=typescript
+```
+
+
+
+#### Working with Custom Templates
+
+For more advanced use cases, you can run `cdk init` with custom templates from Git repositories or NPM packages to:
+
+* Standardize project structure across your organization
+* Include company-specific configurations and best practices
+* Pre-configure common services and patterns
+
+Let’s walk through an example of running `cdk init` in an empty directory called “example-project” with the custom “my-custom-template” TypeScript template below, which we can pull from a Git repository or NPM package:
+
+```
+my-custom-template/
```

Comment:
We ise custom linraries in our cdk apps which are in stored in a folder named `lib` and we put our stacks under `stacks`. Now with the current templating directory structure it forces to use a particular directory structure (if I understood it correctly) which is a bot limiting. Can we do something like

- Use a config file which we can pass to cdk init that tells where to find the app enteypoint template, stack template etc. Or some more cli options. 

- In addition to that I would also need to be able to give arbitray folders which will be copied from the template dir during the init phase. We can 

For example something like below

```bash
cdk init --language=python --template_dir=<path> --app_template_dir=<path> stack_template_dir=<path> --extra_template_dirs=<path,path>
```